libsolidity/formal/CHC.cpp

```diff
@@ -154,9 +155,22 @@ void CHC::endVisit(ContractDefinition const& _contract)
 	smtutil::Expression zeroes(true);
 	for (auto var: stateVariablesIncludingInheritedAndPrivate(_contract))
 		zeroes = zeroes && currentValue(*var) == smt::zeroValue(var->type());
-	addRule(smtutil::Expression::implies(initialConstraints(_contract) && zeroes, predicate(entry)), entry.functor().name);
+
+	smtutil::Expression newAddress = externalCallsIsTrustedMode() ?
+		!state().addressActive(state().thisAddress()) :
+		smtutil::Expression(true);
```

Comment:
This looks complicated with negation, can we make it better?

Reply:
We need the negation of the `addressActive` so, we can add another intermediate variable but not sure it changes much?

Reply:
In principle you could of course do sth like:
```
auto precondition = initialConstraints(_contract) && zeroes;
if (externalCallsIsTrustedMode())
   precondition = move(preconditon) && !state().addressActive(state().thisAddress());
addRule(smtutil::Expression::implies(move(precondition), predicate(entry)), entry.functor().name);
```
or even:
```
auto rule = smtutil::Expression::implies(initialConstraints(_contract) && zeroes, predicate(entry)), entry.functor().name;
if (externalCallsIsTrustedMode())
  rule = smtutil::Expression::implies(!state().addressActive(state().thisAddress()), move(rule));
addRule(move(rule))
```
or logically it should even be the same as
```
auto rule = smtutil::Expression::implies(initialConstraints(_contract) && zeroes && newAddress, predicate(entry));
if (externalCallsIsTrustedMode())
  rule = state().addressActive(state().thisAddress()) || std::move(rule);
addRule(std::move(rule));
```
but that probably makes it worse :-).
I'm fine with just keeping it in any case.